### ISSUE TITLE
J2ME audio example?

### ISSUE THREAD

**Issue by Wellyxc** [SimonKagstrom/cibyl#2]
How to play audio using C and Cibyl?

**Comment by SimonKagstrom** [comment]
wav and mp3 is something I never tried/don't know how to do. All I ever used was simple beeps, which is what e.g., the Sarien port does here

https://github.com/SimonKagstrom/sarien-j2me/blob/master/src/core/sound.c#L429

unfortunately, my memory of Cibyl is getting cloudy by now :-)

**Comment by diaphragm-workplace** [comment]
Time traveler here.
The only mention of that beeping function is [here](https://github.com/SimonKagstrom/cibyl/blob/9b4c303838a81c91e4b0227155be7bc39aa182f5/syscalls/javax.microedition/include/javax/microedition/media.h#L24). "Throws". Throws what? Exceptions? And I couldn't find any other mention of this media interface anywhere else in the tree either.

So, your game is just printing errors to the beat of the music @SimonKagstrom? Absolute cinema!

**Comment by SimonKagstrom** [comment]
@TailsFanLOL no, it's in the link I posted above, or https://github.com/SimonKagstrom/sarien-j2me/blob/a20a3143454dd2e69dde11a086e907012b1372c7/src/core/sound.c#L435

The library you found in Cibyl itself is just a wrapper for the J2ME class Manager, playTone method, which, yes, can throw Java exceptions. So cibyl is just the programming environment and wrappers for Java libraries, whereas the applications themselves that build on Cibyl can play tones etc.

Apparently, some brave people have more recently used Cibyl in some retro computing settings, and at least this repo: https://github.com/r58Playz/celeste-j2me/blob/main/syscalls/celeste/include/audio.h, appears to actually play MP3s, which was the original question here.

**Comment by diaphragm-workplace** [comment]
@SimonKagstrom , well I understand that it is a wrapper, I am just struggling to understand how that header file connects to anything. I am a newbie, sorry.

And, can one merge [the actual wrapper for audio in that repo](https://github.com/r58Playz/celeste-j2me/blob/main/syscalls/celeste/classes/Audio.java)? I mean, this project is kind of dead, but I guess it's never too late. And besides, someone willing to make another port/homebrew would probably not want to look for random addons across obscure repos.

Offtopic honorable mention, someone made [support for Blu-Ray players](https://github.com/sleirsgoevy/bdj-cibyl/blob/master/java/root/org/homebrew/Screen.java) for [a C64 emulator](https://github.com/sleirsgoevy/bdj-frodo) and [Doom](https://github.com/sleirsgoevy/bdj-doom). Probably out of scope for this.

EDIT: not just someone but the one who made the last commit to this repo, whoops

**Comment by SimonKagstrom** [comment]
Yeah, documentation is lacking in Cibyl. The syscall header files are actually parsed to produce the wrappers, and basically follows the pattern `NOPH_<class-name>_<method>.` Objects are passed as the first argument, where applicable, and have a dummy type (int) which is just a handle.

The audio class appears to be a wrapper around some other javax.microedition.media stuff: https://github.com/r58Playz/celeste-j2me/blob/main/syscalls/celeste/classes/Audio.java, so if adding to Cibyl, it would probably make more sense to add the underlying calls from media.


But yes, my cibyl project is mostly dead, as are dumbphones :-). If the dumbphone retro community picks up, I guess I might revisit it some day if I have time and urge to do it. The most pressing feature for me personally would be to modernize the infrastructure, e.g., via docker images, and modernize the compiler.

I integrated a fix which had to be done for the Doom port, which I'm very impressed with. I actually ported Frodo first though, to  my j2me phone back in the day. Was too slow to be useful though, but made for a nifty demo I would say. Sarien was feature complete and playable though.

**Comment by diaphragm-workplace** [comment]
Thanks! Not a fan of containers personally but you do you. Might actually be a good solution if you wanna keep the ancient dependencies...

I think that if one manages to maintain this well again, it will get slightly more popular. And the other way around, if it gets slightly more popular (this year's game jam with exactly 1 submission doesn't count), there would be more consensus/will to maintain it. Chicken/egg problem...

**Comment by SimonKagstrom** [comment]
Yes, I know. However, there's also an effort involved, and in my spare time I tend to do what I find most rewarding. The last decade or so it has been other projects than Cibyl, so I haven't done any real work there in quite a while.

For many years, there were no interest in Cibyl, which makes sense due to the demise of the dumb phone. So even though there has only been a few (it's actually more than one) projects stating to use it lately, that's more than  had happened for many years.

So with that perspective, it might make sense to modernize it. Right now, I'm doing other fun stuff, so not planning on taking it up for some time.

(Updated the comment above, where formatting got lost in the syscall example).

**Comment by diaphragm-workplace** [comment]
My 4 whole braincells thought of an indie idea for J2ME but it would involve a lot of drawing and stuff I don't have time for it. Well, as C418 once said, one could do a million things if they "had the time", but I need school stuff figured out. Long story short, it's a meta narrative game with RPG-like dialogue boxes about the year 2038 problem mass collapsing buildings, a family that owns a ferret with an old man that refuses to give up his old phone, and an evil corporation that made an AI that can run on potato level hardware who hacked all legacy devices (that have cameras) to include it. It so happens that the family ran to a shelter the door to which is controlled electronically, they forgot the phone and the cage outside together, and can't get out of the shelter after the collapsing stopped because some connector got exposed and popped out. The gameplay would be dodging flying text of names of various exceptions, while trying to do some puzzle (haven't decided which) to manipulate pointers and break out of the electronic jail while the game is periodically hinting that the phone's hardware is dying. There would be a boss being the phone's RTC reaching the overflow. The point of all that is to trigger a vibration motor function in the end that will shake the phone and unlock the cage. The ferret will in turn walk curiously and step on the connector, unlocking the door. The whole story will be conveyed with pixel art slideshows. And there will be a demo over activation screen mid game as a little jumpscare. Sounds absurd but I think that's what would make it charming. I even thought of names, "J2meltdown" (the game), "EvilCo" (the corp, duh), "QuackCom Melt" (Qualcomm BREW) and "SimSong" (Siemens, Samsung).

Anyways, we should probably stop talking offtopic. If actual discussion comes in here one day, I don't want the adventurers to scroll too far.

**Comment by SimonKagstrom** [comment]
Sounds like a movie script :-). Manipulating pointers is also what Cibyl brings to Java, so might be a perfect fit!

**Comment by diaphragm-workplace** [comment]
Wow, I tried the Celeste port you linked. Apart from nearly impossible controls, audio doesn't actually play. Neither on my Android powered Scamsung in an emulator or my real J2ME phone. Oh, welp, beeping it is.

**Comment by SimonKagstrom** [comment]
I know nothing about the celeste-j2me port, but just noticed that it had code for MP3 audio. Perhaps @r58Playz has more information!

**Comment by r58Playz** [comment]
None of the emulators I tried could actually play MP3 files, except for J2ME Loader on Android. I don't think any real phones can play multiple sounds at once either, so it's probably not going to play there too. The controls are alright on J2ME Loader and Microemulator, but they could definitely be improved.

**Comment by diaphragm-workplace** [comment]
> except for J2ME Loader on Android

Well, I got news for you: I use that exact emulator.

> I don't think any real phones can play multiple sounds at once either

AFAIK some Sony Ericsson and really late Nokia models support that.

**Comment by diaphragm-workplace** [comment]
Also, I think the beep function (AFAIK it creates and kills a player object immediately) is too slow and it would be like early DOS games with a PC Speaker that only play audio on non-playable parts as to not make everything too laggy. From Oracle docs, apparently there's a format for 1-bit music but that would require actually adding the entire thing to Cybil first. Might as well also add generic playback...
```java
 * "Mary Had A Little Lamb" has "ABAC" structure.
 * Use block to repeat "A" section.
 */
byte tempo = 30; // set tempo to 120 bpm
byte d = 8;      // eighth-note

byte C4 = ToneControl.C4;;
byte D4 = (byte)(C4 + 2); // a whole step
byte E4 = (byte)(C4 + 4); // a major third
byte G4 = (byte)(C4 + 7); // a fifth
byte rest = ToneControl.SILENCE; // rest

byte[] mySequence = {
    ToneControl.VERSION, 1,   // version 1
    ToneControl.TEMPO, tempo, // set tempo
    ToneControl.BLOCK_START, 0,   // start define "A" section
    E4,d, D4,d, C4,d, E4,d,       // content of "A" section
    E4,d, E4,d, E4,d, rest,d,
    ToneControl.BLOCK_END, 0,     // end define "A" section
    ToneControl.PLAY_BLOCK, 0,    // play "A" section
    D4,d, D4,d, D4,d, rest,d,     // play "B" section
    E4,d, G4,d, G4,d, rest,d,
    ToneControl.PLAY_BLOCK, 0,    // repeat "A" section
    D4,d, D4,d, E4,d, D4,d, C4,d  // play "C" section
};

try{
    Player p = Manager.createPlayer(Manager.TONE_DEVICE_LOCATOR);
    p.realize();
    ToneControl c = (ToneControl)p.getControl("ToneControl");
    c.setSequence(mySequence);
    p.start();
} catch (IOException ioe) {
} catch (MediaException me) { }
```
